Parameters lost after KernelProcessor.generateExecutionContext

## Bug Report

### Which version of ShardingSphere did you use?

master - a1da19a3471f1f87ec7e1eb45e25a42322ad249c

### Which project did you use? ShardingSphere-JDBC or ShardingSphere-Proxy?

ShardingSphere-Proxy PostgreSQL.

### Expected behavior

Data inserted.

### Actual behavior

Client:
```
Exception in thread "main" org.postgresql.util.PSQLException: ERROR: No value specified for parameter 1.
	at org.postgresql.core.v3.QueryExecutorImpl.receiveErrorResponse(QueryExecutorImpl.java:2675)
	at org.postgresql.core.v3.QueryExecutorImpl.processResults(QueryExecutorImpl.java:2365)
	at org.postgresql.core.v3.QueryExecutorImpl.execute(QueryExecutorImpl.java:355)
	at org.postgresql.jdbc.PgStatement.executeInternal(PgStatement.java:490)
	at org.postgresql.jdbc.PgStatement.execute(PgStatement.java:408)
	at org.postgresql.jdbc.PgPreparedStatement.executeWithFlags(PgPreparedStatement.java:166)
	at org.postgresql.jdbc.PgPreparedStatement.executeUpdate(PgPreparedStatement.java:134)
```

Proxy:
```
[INFO ] 2022-11-25 16:05:30.051 [Connection-1-ThreadExecutor] ShardingSphere-SQL - Logic SQL: INSERT INTO f (col_f, col_d) VALUES (?::REAL, ?::DOUBLE PRECISION)
[INFO ] 2022-11-25 16:05:30.051 [Connection-1-ThreadExecutor] ShardingSphere-SQL - SQLStatement: PostgreSQLInsertStatement(super=InsertStatement(super=AbstractSQLStatement(parameterCount=2, parameterMarkerSegments=[ParameterMarkerExpressionSegment(startIndex=37, stopIndex=37, parameterMarkerIndex=0, parameterMarkerType=QUESTION, alias=Optional.empty), ParameterMarkerExpressionSegment(startIndex=46, stopIndex=46, parameterMarkerIndex=1, parameterMarkerType=QUESTION, alias=Optional.empty)], commentSegments=[]), table=SimpleTableSegment(tableName=TableNameSegment(startIndex=12, stopIndex=12, identifier=IdentifierValue(value=f, quoteCharacter=NONE)), owner=Optional.empty, alias=Optional.empty), insertColumns=Optional[InsertColumnsSegment(startIndex=14, stopIndex=27, columns=[ColumnSegment(startIndex=15, stopIndex=19, identifier=IdentifierValue(value=col_f, quoteCharacter=NONE), owner=Optional.empty), ColumnSegment(startIndex=22, stopIndex=26, identifier=IdentifierValue(value=col_d, quoteCharacter=NONE), owner=Optional.empty)])], insertSelect=Optional.empty, values=[InsertValuesSegment(startIndex=36, stopIndex=65, values=[CommonExpressionSegment(startIndex=37, stopIndex=43, text=?::REAL), CommonExpressionSegment(startIndex=46, stopIndex=64, text=?::DOUBLE PRECISION)])]), withSegment=Optional.empty, onDuplicateKeyColumnsSegment=null, returningSegment=Optional.empty)
[INFO ] 2022-11-25 16:05:30.052 [Connection-1-ThreadExecutor] ShardingSphere-SQL - Actual SQL: ds_0 ::: INSERT INTO f (col_f, col_d) VALUES (?::REAL, ?::DOUBLE PRECISION)
[ERROR] 2022-11-25 16:05:30.079 [Connection-1-ThreadExecutor] o.a.s.p.f.c.CommandExecutorTask - Exception occur: 
org.postgresql.util.PSQLException: No value specified for parameter 1.
	at org.postgresql.core.v3.SimpleParameterList.checkAllParametersSet(SimpleParameterList.java:284)
	at org.postgresql.core.v3.QueryExecutorImpl.execute(QueryExecutorImpl.java:340)
	at org.postgresql.jdbc.PgStatement.executeInternal(PgStatement.java:490)
	at org.postgresql.jdbc.PgStatement.execute(PgStatement.java:408)
	at org.postgresql.jdbc.PgPreparedStatement.executeWithFlags(PgPreparedStatement.java:181)
	at org.postgresql.jdbc.PgPreparedStatement.execute(PgPreparedStatement.java:170)
	at com.zaxxer.hikari.pool.ProxyPreparedStatement.execute(ProxyPreparedStatement.java:44)
	at com.zaxxer.hikari.pool.HikariProxyPreparedStatement.execute(HikariProxyPreparedStatement.java)
	at org.apache.shardingsphere.proxy.backend.communication.jdbc.executor.callback.impl.ProxyPreparedStatementExecutorCallback.execute(ProxyPreparedStatementExecutorCallback.java:43)
	at org.apache.shardingsphere.proxy.backend.communication.jdbc.executor.callback.ProxyJDBCExecutorCallback.executeSQL(ProxyJDBCExecutorCallback.java:75)
	at org.apache.shardingsphere.proxy.backend.communication.jdbc.executor.callback.ProxyJDBCExecutorCallback.executeSQL(ProxyJDBCExecutorCallback.java:68)
	at org.apache.shardingsphere.proxy.backend.communication.jdbc.executor.callback.ProxyJDBCExecutorCallback.executeSQL(ProxyJDBCExecutorCallback.java:45)
	at org.apache.shardingsphere.infra.executor.sql.execute.engine.driver.jdbc.JDBCExecutorCallback.execute(JDBCExecutorCallback.java:90)
	at org.apache.shardingsphere.infra.executor.sql.execute.engine.driver.jdbc.JDBCExecutorCallback.execute(JDBCExecutorCallback.java:69)
	at org.apache.shardingsphere.infra.executor.kernel.ExecutorEngine.syncExecute(ExecutorEngine.java:135)
	at org.apache.shardingsphere.infra.executor.kernel.ExecutorEngine.parallelExecute(ExecutorEngine.java:131)
	at org.apache.shardingsphere.infra.executor.kernel.ExecutorEngine.execute(ExecutorEngine.java:116)
	at org.apache.shardingsphere.infra.executor.sql.execute.engine.driver.jdbc.JDBCExecutor.execute(JDBCExecutor.java:67)
	at org.apache.shardingsphere.proxy.backend.communication.jdbc.executor.ProxyJDBCExecutor.execute(ProxyJDBCExecutor.java:75)
	at org.apache.shardingsphere.proxy.backend.communication.ProxySQLExecutor.useDriverToExecute(ProxySQLExecutor.java:229)
	at org.apache.shardingsphere.proxy.backend.communication.ProxySQLExecutor.doExecute(ProxySQLExecutor.java:186)
	at org.apache.shardingsphere.proxy.backend.communication.ProxySQLExecutor.execute(ProxySQLExecutor.java:151)
	at org.apache.shardingsphere.proxy.backend.communication.jdbc.JDBCDatabaseCommunicationEngine.execute(JDBCDatabaseCommunicationEngine.java:131)
	at org.apache.shardingsphere.proxy.frontend.postgresql.command.query.extended.JDBCPortal.bind(JDBCPortal.java:101)
	at org.apache.shardingsphere.proxy.frontend.postgresql.command.query.extended.bind.PostgreSQLComBindExecutor.execute(PostgreSQLComBindExecutor.java:53)
	at org.apache.shardingsphere.proxy.frontend.postgresql.command.query.extended.PostgreSQLAggregatedCommandExecutor.execute(PostgreSQLAggregatedCommandExecutor.java:41)
	at org.apache.shardingsphere.proxy.frontend.command.CommandExecutorTask.executeCommand(CommandExecutorTask.java:111)
	at org.apache.shardingsphere.proxy.frontend.command.CommandExecutorTask.run(CommandExecutorTask.java:78)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:833)
```

### Reason analyze (If you can)

![image](https://user-images.githubusercontent.com/20503072/203930697-cc219a7b-d460-4416-87c3-8e930f4bc556.png)

### Steps to reproduce the behavior, such as: SQL to execute, sharding rule configuration, when exception occur etc.

No rule configured.
```yaml
schemaName: postgres
dataSources:
  ds_0:
    url: jdbc:postgresql://127.0.0.1:5432/postgres?prepareThreshold=1
    username: postgres
    password: postgres
    connectionTimeoutMilliseconds: 3000
    idleTimeoutMilliseconds: 60000
    maxLifetimeMilliseconds: 1800000
    maxPoolSize: 32
    minPoolSize: 0
```

### Example codes for reproduce this issue (such as a github link).

```sql
create table f (col_f real, col_d double precision);
```

```java
        try (PreparedStatement preparedStatement = connection.prepareStatement("INSERT INTO f (col_f, col_d) VALUES (?::REAL, ?::DOUBLE PRECISION)")) {
            preparedStatement.setObject(1, "Infinity");
            preparedStatement.setObject(2, "Infinity");
            preparedStatement.executeUpdate();
        }

```

About the type cast of PostgreSQL: https://www.postgresql.org/docs/15/sql-expressions.html#SQL-SYNTAX-TYPE-CASTS

--- a/sql-parser/dialect/opengauss/src/main/java/org/apache/shardingsphere/sql/parser/opengauss/visitor/statement/impl/OpenGaussStatementSQLVisitor.java
+++ b/sql-parser/dialect/opengauss/src/main/java/org/apache/shardingsphere/sql/parser/opengauss/visitor/statement/impl/OpenGaussStatementSQLVisitor.java
@@ -299,6 +299,10 @@ public abstract class OpenGaussStatementSQLVisitor extends OpenGaussStatementBas
         if (null != ctx.cExpr()) {
             return visit(ctx.cExpr());
         }
+        if (null != ctx.TYPE_CAST_()) {
+            return new BinaryOperationExpression(ctx.start.getStartIndex(), ctx.stop.getStopIndex(), (ExpressionSegment) visit(ctx.aExpr(0)),
+                    new CommonExpressionSegment(ctx.typeName().start.getStartIndex(), ctx.typeName().stop.getStopIndex(), ctx.typeName().getText()), ctx.TYPE_CAST_().getText(), ctx.getText());
+        }
         if (null != ctx.BETWEEN()) {
             return createBetweenSegment(ctx);
         }

--- a/sql-parser/dialect/postgresql/src/main/java/org/apache/shardingsphere/sql/parser/postgresql/visitor/statement/impl/PostgreSQLStatementSQLVisitor.java
+++ b/sql-parser/dialect/postgresql/src/main/java/org/apache/shardingsphere/sql/parser/postgresql/visitor/statement/impl/PostgreSQLStatementSQLVisitor.java
@@ -299,6 +299,10 @@ public abstract class PostgreSQLStatementSQLVisitor extends PostgreSQLStatementP
         if (null != ctx.cExpr()) {
             return visit(ctx.cExpr());
         }
+        if (null != ctx.TYPE_CAST_()) {
+            return new BinaryOperationExpression(ctx.start.getStartIndex(), ctx.stop.getStopIndex(), (ExpressionSegment) visit(ctx.aExpr(0)),
+                    new CommonExpressionSegment(ctx.typeName().start.getStartIndex(), ctx.typeName().stop.getStopIndex(), ctx.typeName().getText()), ctx.TYPE_CAST_().getText(), ctx.getText());
+        }
         if (null != ctx.BETWEEN()) {
             return createBetweenSegment(ctx);
         }

--- a/test/e2e/suite/src/test/resources/cases/dml/dml-integration-test-cases.xml
+++ b/test/e2e/suite/src/test/resources/cases/dml/dml-integration-test-cases.xml
@@ -310,12 +310,9 @@
         <assertion parameters="3:int, -1:long, -1:int, -1:int, -1:smallint, -1:tinyint" expected-data-file="insert_negative_values_into_single_table_integer.xml" />
     </test-case>
     
-    <!--    TODO Fixme https://github.com/apache/shardingsphere/issues/22407 -->
-    <!--    <test-case sql="INSERT INTO t_data_type_floating_point (id, col_float, col_double) VALUES (?, ?::REAL, ?::DOUBLE PRECISION)" db-types="PostgreSQL,openGauss" scenario-types="passthrough">-->
-    <!--        <assertion parameters="1:int, Infinity:String, Infinity:String" expected-data-file="insert_positive_infinity_floating_point_values.xml"/>-->
-    <!--        <assertion parameters="2:int, -Infinity:String, -Infinity:String" expected-data-file="insert_negative_infinity_floating_point_values.xml"/>-->
-    <!--    </test-case>-->
-    <test-case sql="INSERT INTO t_data_type_floating_point (id, col_float, col_double) VALUES (?, ?, ?)" db-types="PostgreSQL,openGauss" scenario-types="passthrough">
+    <test-case sql="INSERT INTO t_data_type_floating_point (id, col_float, col_double) VALUES (?, ?::REAL, ?::DOUBLE PRECISION)" db-types="PostgreSQL,openGauss" scenario-types="passthrough">
+        <assertion parameters="1:int, Infinity:String, Infinity:String" expected-data-file="insert_positive_infinity_floating_point_values.xml"/>
+        <assertion parameters="2:int, -Infinity:String, -Infinity:String" expected-data-file="insert_negative_infinity_floating_point_values.xml"/>
         <assertion parameters="3:int, 3.4028235E38:float, 1.7976931348623157E308:double" expected-data-file="insert_max_floating_point_values.xml" />
         <assertion parameters="4:int, 1.4E-45:float, 4.9E-324:double" expected-data-file="insert_min_floating_point_values.xml" />
     </test-case>

--- a/test/it/parser/src/main/resources/case/dml/insert.xml
+++ b/test/it/parser/src/main/resources/case/dml/insert.xml
@@ -41,6 +41,45 @@
             </value>
         </values>
     </insert>
+    
+    <insert sql-case-id="insert_with_historical_type_cast_syntax" parameters="1, 1, 'init'">
+        <table name="t_order" start-index="12" stop-index="18" />
+        <columns start-index="20" stop-index="46">
+            <column name="order_id" start-index="21" stop-index="28" />
+            <column name="user_id" start-index="31" stop-index="37" />
+            <column name="status" start-index="40" stop-index="45" />
+        </columns>
+        <values>
+            <value>
+                <assignment-value>
+                    <parameter-marker-expression parameter-index="0" start-index="56" stop-index="56" />
+                    <literal-expression value="1" start-index="56" stop-index="56" />
+                </assignment-value>
+                <assignment-value>
+                    <binary-operation-expression start-index="59" stop-index="65">
+                        <left>
+                            <parameter-marker-expression parameter-index="2" start-index="59" stop-index="59" />
+                        </left>
+                        <right>
+                            <common-expression text="int4" start-index="62" stop-index="65" />
+                        </right>
+                        <operator>::</operator>
+                    </binary-operation-expression>
+                </assignment-value>
+                <assignment-value>
+                    <binary-operation-expression start-index="68" stop-index="74">
+                        <left>
+                            <parameter-marker-expression parameter-index="3" start-index="68" stop-index="68" />
+                        </left>
+                        <right>
+                            <common-expression text="text" start-index="71" stop-index="74" />
+                        </right>
+                        <operator>::</operator>
+                    </binary-operation-expression>
+                </assignment-value>
+            </value>
+        </values>
+    </insert>
 
     <insert sql-case-id="insert_with_now_function" parameters="1, 1, 'init'">
         <table name="t_order_item" start-index="12" stop-index="23" />

--- a/test/it/parser/src/main/resources/case/dml/select.xml
+++ b/test/it/parser/src/main/resources/case/dml/select.xml
@@ -1960,7 +1960,15 @@
                                             </left>
                                             <operator>=</operator>
                                             <right>
-                                                <common-expression text="'{&quot;rule2&quot;:&quot;null2&quot;}'::jsonb" start-index="35" stop-index="60" />
+                                                <binary-operation-expression start-index="35" stop-index="60">
+                                                    <left>
+                                                        <literal-expression value="{&quot;rule2&quot;:&quot;null2&quot;}" start-index="35" stop-index="53" />
+                                                    </left>
+                                                    <operator>::</operator>
+                                                    <right>
+                                                        <common-expression text="jsonb" start-index="56" stop-index="60" />
+                                                    </right>
+                                                </binary-operation-expression>
                                                 <parameter-marker-expression parameter-index="0" start-index="35" stop-index="42" />
                                             </right>
                                         </binary-operation-expression>
@@ -2039,11 +2047,27 @@
                                             <right>
                                                 <list-expression start-index="36" stop-index="55" literal-stop-index="91">
                                                     <items>
-                                                        <common-expression text="'{&quot;rule2&quot;:&quot;null2&quot;}'::jsonb" start-index="37" stop-index="62" />
+                                                        <binary-operation-expression start-index="37" stop-index="62">
+                                                            <left>
+                                                                <literal-expression value="{&quot;rule2&quot;:&quot;null2&quot;}" start-index="37" stop-index="55" />
+                                                            </left>
+                                                            <operator>::</operator>
+                                                            <right>
+                                                                <common-expression text="jsonb" start-index="58" stop-index="62" />
+                                                            </right>
+                                                        </binary-operation-expression>
                                                         <parameter-marker-expression parameter-index="0" start-index="37" stop-index="44" />
                                                     </items>
                                                     <items>
-                                                        <common-expression text="'{&quot;rule3&quot;:&quot;null3&quot;}'::jsonb" start-index="65" stop-index="90" />
+                                                        <binary-operation-expression start-index="65" stop-index="90">
+                                                            <left>
+                                                                <literal-expression value="{&quot;rule3&quot;:&quot;null3&quot;}" start-index="65" stop-index="83" />
+                                                            </left>
+                                                            <operator>::</operator>
+                                                            <right>
+                                                                <common-expression text="jsonb" start-index="86" stop-index="90" />
+                                                            </right>
+                                                        </binary-operation-expression>
                                                         <parameter-marker-expression parameter-index="1" start-index="47" stop-index="54" />
                                                     </items>
                                                 </list-expression>
@@ -2132,7 +2156,15 @@
                                         </binary-operation-expression>
                                     </between-expr>
                                     <and-expr>
-                                        <common-expression text="'{&quot;rule3&quot;:&quot;null3&quot;}'::jsonb" start-index="72" stop-index="97" />
+                                        <binary-operation-expression start-index="72" stop-index="97">
+                                            <left>
+                                                <literal-expression value="{&quot;rule3&quot;:&quot;null3&quot;}" start-index="72" stop-index="90" />
+                                            </left>
+                                            <operator>::</operator>
+                                            <right>
+                                                <common-expression text="jsonb" start-index="93" stop-index="97" />
+                                            </right>
+                                        </binary-operation-expression>
                                         <parameter-marker-expression parameter-index="1" start-index="54" stop-index="61" />
                                     </and-expr>
                                 </between-expression>

--- a/test/it/parser/src/main/resources/sql/supported/dml/insert.xml
+++ b/test/it/parser/src/main/resources/sql/supported/dml/insert.xml
@@ -18,6 +18,7 @@
 
 <sql-cases>
     <sql-case id="insert_with_all_placeholders" value="INSERT INTO t_order (order_id, user_id, status) VALUES (?, ?, ?)" />
+    <sql-case id="insert_with_historical_type_cast_syntax" value="INSERT INTO t_order (order_id, user_id, status) VALUES (?, ?::int4, ?::text)" db-types="PostgreSQL,openGauss" />
     <sql-case id="insert_with_now_function" value="INSERT INTO t_order_item (item_id, order_id, user_id, status, creation_date) VALUES (?, ?, ?, 'insert', now())" db-types="MySQL" />
     <sql-case id="insert_without_parameters" value="INSERT INTO t_order (order_id, user_id, status) VALUES (1, 1, 'insert')" />
     <sql-case id="insert_with_special_characters" value="INSERT INTO `t_order` (`order_id`, `user_id`, `status`) VALUES (1, 1, 'insert')" db-types="MySQL" />

